lib/SILOptimizer/Mandatory/Differentiation.cpp

```diff
@@ -1062,10 +1062,29 @@ void ADContext::emitNondifferentiabilityError(SILInstruction *inst,
   // that is not associated with any source location, we emit a diagnostic at
   // the instruction source location.
   case DifferentiationInvoker::Kind::AutoDiffFunctionInst:
-  case DifferentiationInvoker::Kind::SILDifferentiableAttribute:
+  case DifferentiationInvoker::Kind::SILDifferentiableAttribute: {
+    auto attr = invoker.getSILDifferentiableAttribute();
     diagnose(opLoc,
              diag.getValueOr(diag::autodiff_expression_is_not_differentiable));
+    if (auto declContext = attr.second->getDeclContext()) {
+      if (auto fnDecl = declContext->getInnermostDeclarationDeclContext()) {
+        if (auto *diffAttr =
+                fnDecl->getAttrs().getAttribute<DifferentiableAttr>()) {
+          // FIXME: The highlight doesn't work when the attr is not on the same
+          // line as the function.
+          diagnose(diffAttr->getLocation(),
+                   diag::autodiff_function_not_differentiable)
+              .highlight(attr.second->getLocation().getSourceLoc());
```

Comment:
We need `DifferentiableAttr` to store the location of the `)`. I'd suggest adding a `rParenLoc` to `DifferentiableAttr` in a follow-up patch.

`@differentiable` rarely appears on the same line as the function declaration. Does this highlighting logic result in misleading behavior if `@differentiable` appears on top of the function declaration?